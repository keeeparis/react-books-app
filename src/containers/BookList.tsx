import React from 'react'
import useBooks from '../hooks/useBooks'
import { useAppSelector } from '../redux/hooks/hooks'
import {
  selectInput,
  selectStartIndex,
  selectSkip,
} from '../redux/mainFeature/mainSlice'
import { useGetBooksQuery } from '../redux/services/books'

const BookList = () => {
  const startIndex = useAppSelector(selectStartIndex)
  const input = useAppSelector(selectInput)
  const isSkip = useAppSelector(selectSkip)

  const { data, isLoading, isFetching, isSuccess, isUninitialized, isError } =
    useGetBooksQuery({ input, startIndex }, { skip: isSkip })

  const { books } = useBooks(data, startIndex)

  return (
    <div>
      {books.length
        ? books.map((book: any) => (
            <div key={`${book.id} + ${book.etag}`}>{book.volumeInfo.title}</div>
          ))
        : null}

      {isFetching ? (
        <div>LOADING...</div>
      ) : isUninitialized ? (
        <div>Давай, смелее!</div>
      ) : isError ? (
        <div>Произошла Ошибка</div>
      ) : (
        !books.length && (
          <div>Мы не смогли найти ни одной книги по вашему запросу.</div>
        )
      )}
    </div>
  )
}

export default BookList
