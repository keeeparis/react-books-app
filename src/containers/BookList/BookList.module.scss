.list {
  display: grid;
  grid-template-columns: repeat(4, 300px);
  gap: 15px;
  grid-auto-rows: minmax(400px, auto);
}

@media (max-width: 1276px) {
  .list {
    grid-template-columns: repeat(3, 300px);
  }
}

@media (max-width: 973px) {
  .list {
    grid-template-columns: repeat(2, 300px);
  }
}

@media (max-width: 660px) {
  .list {
    grid-template-columns: 1fr;
  }
}
