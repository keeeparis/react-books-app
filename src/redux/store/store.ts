import { configureStore } from '@reduxjs/toolkit'
import { setupListeners } from '@reduxjs/toolkit/dist/query'

import mainReducer from '../mainFeature/mainSlice'

import { booksApi } from '../services/books'

export const store = configureStore({
  reducer: {
    main: mainReducer,
    [booksApi.reducerPath]: booksApi.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(booksApi.middleware),
})

// setupListeners(store.dispatch)

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
