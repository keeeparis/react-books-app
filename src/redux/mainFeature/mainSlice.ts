import { createSlice } from '@reduxjs/toolkit'
import { RootState } from '../store/store'

interface mainState {
  page: number
  skip: boolean
  input: string
}

const initialState: mainState = {
  page: 1,
  skip: true,
  input: '',
}

export const mainSlice = createSlice({
  name: 'main',
  initialState,
  reducers: {
    updatePage: (state, action) => {
      state.page += action.payload
    },
    updateSkip: (state, action) => {
      state.skip = action.payload
    },
    updateInput: (state, action) => {
      state.input = action.payload
    },
  },
})

export const { updatePage, updateSkip, updateInput } = mainSlice.actions

export default mainSlice.reducer

export const selectPage = (state: RootState) => state.main.page
export const selectSkip = (state: RootState) => state.main.skip
export const selectInput = (state: RootState) => state.main.input
