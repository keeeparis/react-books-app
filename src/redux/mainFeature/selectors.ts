import { createSelector } from '@reduxjs/toolkit'

import { BOOKS_PER_PAGE } from '../../pages/App'
import { RootState } from '../store/store'

/* Select */
export const selectStartIndex = (state: RootState) => state.main.startIndex
export const selectSkip = (state: RootState) => state.main.skip
export const selectInput = (state: RootState) => state.main.input
export const selectAllBooks = (state: RootState) => state.main.books
export const selectCategory = (state: RootState) => state.main.category
export const selectSorting = (state: RootState) => state.main.sorting
export const selectTotalItems = (state: RootState) => state.main.totalItems

export const selectIsMoreResults = (state: RootState) =>
  state.main.totalItems > state.main.startIndex + BOOKS_PER_PAGE

/* Create Selectors */
export const selectIsAnyBooks = createSelector(selectAllBooks, (books) =>
  books ? !!books.length : 0
)

export const selectBookById = createSelector(
  [selectAllBooks, (state, id: string) => id],
  (books, id) => {
    return books.filter((book) => book.id === id)[0]
  }
)
