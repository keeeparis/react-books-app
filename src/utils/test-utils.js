import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'
import 'regenerator-runtime/runtime'
import { Provider } from 'react-redux'
import { store } from '../redux/store/store'
import { MemoryRouter } from 'react-router-dom'
import '../i18n/config'

const AllTheProviders = ({ children }) => {
  return (
    <Provider store={store}>
      <MemoryRouter>{children}</MemoryRouter>
    </Provider>
  )
}

const customRender = (ui, options) => {
  return render(ui, { wrapper: AllTheProviders, ...options })
}

export * from '@testing-library/react'
export * from '@testing-library/jest-dom'
export { userEvent }
export * from 'regenerator-runtime/runtime'
export * from '../i18n/config'

export { customRender as render }
