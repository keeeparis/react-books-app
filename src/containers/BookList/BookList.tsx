import React from 'react'
import { useAppSelector } from '../../redux/hooks/hooks'
import { selectCategorizedBooks } from '../../redux/mainFeature/selectors'
import BookCard from '../../components/BookCard'
import styles from './BookList.module.scss'

const BookList = () => {
  const books = useAppSelector(selectCategorizedBooks)

  return (
    <div className={styles.list}>
      {books.length
        ? books.map((book) => (
            <BookCard
              key={`${book.id} + ${book.etag}`}
              book={book.volumeInfo}
              id={book.id}
            />
          ))
        : null}
    </div>
  )
}

export default BookList
