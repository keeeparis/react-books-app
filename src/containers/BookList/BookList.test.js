import { act } from 'react-test-renderer'
import { render, screen, fireEvent, userEvent } from 'test-utils'
import { addBooks } from '../../redux/mainFeature/mainSlice'
import { store } from '../../redux/store/store'
import BookList from './BookList'

const books = [
  {
    etag: '1',
    id: '1',
    volumeInfo: {
      authors: ['Conan', 'Boyle'],
      imageLinks: {
        smallThumbnail: 'https://url.com',
        thumbnail: 'https://url.com',
      },
      categories: ['Horror', 'Fiction'],
      pageCount: 30,
      description: 'Some text in description',
      language: 'ru',
      title: 'VERY GOOD TITLE',
      publishedDate: '1999',
      averageRating: 10,
    },
  },
  {
    etag: '2',
    id: '2',
    volumeInfo: {
      authors: ['Kayle', 'Dock'],
      imageLinks: {
        smallThumbnail: 'https://url2.com',
        thumbnail: 'https://url2.com',
      },
      categories: ['Science', 'Melody'],
      pageCount: 900,
      description: 'Another text in description',
      language: 'es',
      title: 'This is THE title',
      publishedDate: '2380',
      averageRating: 2,
    },
  },
]

it('test BookList Component', () => {
  const { asFragment } = render(<BookList />)

  expect(asFragment()).toMatchSnapshot()

  act(() => {
    store.dispatch(addBooks(books))
  })

  expect(screen.getByText(/VERY GOOD TITLE/)).toBeInTheDocument()
  expect(screen.getByText(/This is THE title/)).toBeInTheDocument()
})
