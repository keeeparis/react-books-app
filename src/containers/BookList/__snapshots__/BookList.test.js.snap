// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test BookList Component 1`] = `
<DocumentFragment>
  <div
    class="list"
  />
</DocumentFragment>
`;
